src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -91,6 +106,94 @@ public EETypeNode(NodeFactory factory, TypeDesc type)
             static TypeDesc WithoutParameterizeTypes(TypeDesc t) => t is ParameterizedType pt ? WithoutParameterizeTypes(pt.ParameterType) : t;
         }
 
+        private static VirtualMethodAnalysisFlags AnalyzeVirtualMethods(TypeDesc type)
+        {
+            VirtualMethodAnalysisFlags result = default;
+
+            // Interface EETypes not relevant to virtual method analysis at this time.
+            if (type.IsInterface)
+                return result;
+
+            DefType defType = type.GetClosestDefType();
+
+            foreach (MethodDesc method in defType.GetAllVirtualMethods())
+            {
+                // First, check if this type has any GVM that overrides a GVM on a parent type. If that's the case, this makes
+                // the current type interesting for GVM analysis (i.e. instantiate its overriding GVMs for existing GVMDependenciesNodes
+                // of the instantiated GVM on the parent types).
+                if (method.HasInstantiation)
+                {
+                    result |= VirtualMethodAnalysisFlags.NeedsGvmEntries;
+
+                    MethodDesc slotDecl = MetadataVirtualMethodAlgorithm.FindSlotDefiningMethodForVirtualMethod(method);
+                    if (slotDecl != method)
+                        result |= VirtualMethodAnalysisFlags.InterestingForDynamicDependencies;
+                }
+
+                // Early out if we set all the flags we could have set
+                if ((result & VirtualMethodAnalysisFlags.AllFlags) == VirtualMethodAnalysisFlags.AllFlags)
+                    return result;
+            }
+
+            //
+            // Check if the type implements any interface, where the method implementations could be on
+            // base types.
+            // Example:
+            //      interface IFace {
+            //          void IFaceGVMethod<U>();
+            //      }
+            //      class BaseClass {
+            //          public virtual void IFaceGVMethod<U>() { ... }
+            //      }
+            //      public class DerivedClass : BaseClass, IFace { }
+            //
+            foreach (DefType interfaceImpl in defType.RuntimeInterfaces)
+            {
+                foreach (MethodDesc method in interfaceImpl.GetAllVirtualMethods())
+                {
+                    if (!method.HasInstantiation)
+                        continue;
+
+                    // We found a GVM on one of the implemented interfaces. Find if the type implements this method.
+                    // (Note, do this comparison against the generic definition of the method, not the specific method instantiation
+                    MethodDesc slotDecl = method.Signature.IsStatic ?
+                        defType.ResolveInterfaceMethodToStaticVirtualMethodOnType(method)
```

Comment:
We might be able to simplify to one method once all type system consumers can deal with static methods. crossgen2 currently still doesn't handle static virtuals and I'm not sure if changing this could break it.

(This is a copy from one of the places that I'm unifying - not new code).